031-set-pipeline-step/proposal.md

```diff
@@ -0,0 +1,74 @@
+# Summary
+
+This RFC proposes a new `set_pipeline` step type for configuring pipelines within a build plan.
+
+
+# Motivation
+
+## Short-term motivation
+
+Lots of folks are already using the [`concourse-pipeline` resource](https://github.com/concourse/concourse-pipeline-resource), however the resource has two fatal flaws:
+
+* Users have to configure a local auth user and pass it to the resource definition.
+* The resource is versioned independently of the user's Concourse, meaning the `fly` version won't always be in sync. The resource makes an attempt to resolve this by doing a `sync` after logging in, but this is a pretty clunky problem regardless.
+
+If we had native support for a `set_pipeline` step, both of these problems would go away.
+
+## Long-term motivation
+
+By having a `set_pipeline` step in the build plan, we can start to improve Concourse's story around automating the full CI stack for projects of all sizes. Users can start to trust that pipelines are always configured via CI, and they can go over the build history to see who changed what and when.
+
+Later RFCs (namely, 'projects' and 'instanced pipelines') will build on this idea to provide a truly continuous workflow for automating pipelines - including their automatic archival when they're no longer needed, in the case of instanced pipelines.
+
+
+# Proposal
+
+Using the step would look something like this:
+
+```yaml
+plan:
+- get: ci
+- set_pipeline: concourse
+  file: ci/pipelines/concourse.yml
+```
+
+The `x` in `set_pipeline: x` is the pipeline name, and `file:` would be used to specify the pipeline config.
+
+The pipeline would be configured within whichever team the build execution belongs to.
+
+The pipeline would be automatically unpaused, as opposed to `fly set-pipeline` which pauses pipelines by default. The assumption here is that if you're automating `set_pipeline` you're not just kicking the tires and can probably trust the pipelines that you're configuring are correct, at least enough to have made it into version control.
+
+## `((vars))` support
+
+Additionally, we should support `vars` (as in `fly set-pipeline -y`) and `vars_files` (i.e. `fly set-pipeline -l`):
```

Comment:
I think this is the most appropriate place to make this comment. I am simple and this is complicated, so please bear with me as I try to explain.

Have you considered use cases where the `set-pipeline` step needs to _use_ vars from the `vars_files`?

We chain together pipelines (all from the same `pipeline.yml` from the same repo, albeit with different versions) that represent different environments (ie dev, staging, prod). The `get`s of important resources that we want to test and promote together have parameterised `version` fields.

1. In the 'dev' pipeline, `get: { version:` is `latest` for all important resources we're integrating, **including the CI resource** that the pipeline is set from
1. Dev pipeline runs, integration tests run.
1. At the end of the pipeline, we use [stopover](https://github.com/EngineerBetter/stopover) to emit a YAML file of the versions of all resources we want to promote into the next environment.
1. Versions YAML file is uploaded to S3
1. The 'staging' pipeline was already configured to look at the S3 bucket, and thus a run of the pipeline is triggered. Notably, at this stage **the pipeline is configured according to an earlier version of the CI resource** (the previous one that passed 'dev').
1. The `set-pipeline` job runs, using the versions file as a `vars_file`. Because all the `get`s' `version`s are parameterised, this effectively constrains versions of resources available to this pipeline. At the point `set-pipeline` is invoked the runtime state of the pipeline is such that **the version of the CI resource is still set to that of the last run**, because the new vars haven't been applied - that won't happen until _after_ `set-pipeline` has succeeded.
1. The runtime state of the pipeline is now that the `get` of the CI resource is constrained to the newer version that passed 'dev'. 
1. We then need to run `set-pipeline` _again_, because only _now_ is the newer version that passed `dev` available to this pipeline
1. Pipeline runs as normal
1. Repeat until production glory

We use this pattern with multiple customers for their pretty important infrastructure stuff. In case anyone's wondering, the benefit it is *guarantees* that no untested versions can sneak into higher environments.

So I guess I have two questions, and because I'm not very clever the answer might already be clear based on the RFC, so apologies if you have to spell this out:

1. **Is the approach above compatible with the new proposal?** My understanding is that is _should_ be, as long as there's no limit on where the new `set_pipeline` step can appear, and how often.
1. **Could the new functionality do this atomically**? Being able to avoid the `set-pipeline` 'double-tap' would make things a lot more clear for us and our customers.

You have my contact details @vito, so let me know if I can help by explaining in more detail on a call or similar.

Reply:
I don't think I have the full picture here as to why the double-tap is necessary but I don't see this proposal forbidding any part of it so long as there's still a way to configure vars from a file. :thinking: